Fix CustomField Bug

jdrew82 opened the pull request:
This PR is to fix the bug detailed in #274. I believe this is around CustomFields that have dashes in the key when it should be underscores instead. There is a migration included to update any user that might have the old CustomFields and migrate them to the new ones.

glennmatthews commented on docs/user/integrations/ipfabric.md:
```diff
@@ -118,7 +118,7 @@ If you would like to change the default status change value, ensure you provide
 ![Safe Delete](../../images/ipfabric-safe-delete.png)
 
-An example object that's been modified by SSoT App and tagged as `ssot-safe-delete` and `ssot-synced-from-ipfabric`. Notice the Status and child object, IPAddress has also changed to Deprecated and, it's status changed and tagged as well.
+An example object that's been modified by SSoT App and tagged as `ssot-safe-delete` and `ssot_last_synchronized`. Notice the Status and child object, IPAddress has also changed to Deprecated and, it's status changed and tagged as well.
```
Does `ssot-safe-delete` need to be changed too? What about updating the screenshot?

jdrew82 replied:
I saw the `ssot-safe-delete` as a Tag not a CustomField so left it alone.

glennmatthews replied:
Ah then maybe just a wording update and screenshot update? The screenshot shows both as tags?

jdrew82 replied:
There is a Tag on the objects `SSoT Synced from IPFabric` and also a CustomField identifying IPFabric as the System of Record. The Tag is used for filtering what's imported depending upon the Job form option selection. I've updated the docs to clarify what's specifically being spoken about.

glennmatthews commented on nautobot_ssot/integrations/ipfabric/diffsync/diffsync_models.py:
```diff
@@ -129,7 +129,7 @@ def update(self, attrs):
             if device_tags.exists():
                 location.tags.remove(safe_delete_tag)
-        tonb_nbutils.tag_object(nautobot_object=location, custom_field="ssot-synced-from-ipfabric")
+        tonb_nbutils.tag_object(nautobot_object=location, custom_field="ssot_last_synchronized")
```
Seeing how many places this is needed, it might be appropriate to define a constant for `ssot_last_synchronized` somewhere?

jdrew82 replied:
I'm not sure I follow. This is meant to be a date record of the last time the particular object was updated from the System of Record.

glennmatthews replied:
I meant a constant like `LAST_SYNCHRONIZED_CF_NAME = "ssot_last_synchronized"` and use that rather than string literals all over the place.

jdrew82 replied:
Got it, can do.

glennmatthews commented on nautobot_ssot/integrations/servicenow/diffsync/adapter_nautobot.py:
```diff
@@ -127,7 +127,7 @@ def load(self):
     def tag_involved_objects(self, target):
         """Tag all objects that were successfully synced to the target."""
-        # The ssot-synced-to-servicenow tag *should* have been created automatically during plugin installation
+        # The servicenow_last_synchronized tag *should* have been created automatically during plugin installation
```
Why isn't this also using the `ssot_last_synchronized` and `system_of_record` cfs?

glennmatthews replied:
Hm, then maybe something like `last_synced_from_ssot` instead of `ssot_last_synchronized` would be clearer?

jdrew82 replied:
I was using `ssot_last_synchronized` as that's what I've used in other projects but that suggestion is slightly shorter and yes I can see it being clearer.

glennmatthews commented on nautobot_ssot/migrations/0007_remove_dashed_custom_fields.py:
perhaps `replace_dashed_custom_fields`?

jdrew82 replied:
Done.

glennmatthews commented on nautobot_ssot/migrations/0007_remove_dashed_custom_fields.py:
```diff
@@ -0,0 +1,50 @@
+            custom_field.save()
+
+    for ct in ContentType.objects.filter(FeatureQuery("custom_fields").get_query()):
+        relevant_custom_fields = CustomField.objects.filter(content_types=ct)
```
Perhaps narrow this down further with something like `.filter(content_types=ct, key__in=CF_KEY_CHANGE_MAP.values())` so that we are just iterating the models that these specific CFs apply to rather than all models that have *any* CF defined?

jdrew82 replied:
Good idea, done!

glennmatthews commented on nautobot_ssot/migrations/0007_remove_dashed_custom_fields.py:
```diff
@@ -0,0 +1,50 @@
+        model = apps.get_model(ct.app_label, ct.model)
+        cf_list = []
+        for instance in model.objects.all():
+            new_custom_field_data = {}
+            for cf in relevant_custom_fields:
+                new_custom_field_data[cf.key] = instance._custom_field_data.pop(CF_KEY_CHANGE_MAP.get(cf.key), None)
+            instance._custom_field_data = new_custom_field_data
```
This probably needs to be reworked since I think as written it's going to throw away any custom field value that *isn't* in CF_KEY_CHANGE_MAP?

jdrew82 replied:
AH, this was based off the other referenced migration.

glennmatthews replied:
Yeah, the other migration was explicitly processing *all* CFs and storing them into `CF_KEY_TO_NAME`. Here we're just operating on a subset.

jdrew82 replied:
Got it.

alhogan commented:
Why `ssot_last_synchronized` and `servicenow_last_synchronized` instead both being updated to `ssot_last_synchronized` or including the integration name in both?

alhogan commented:
Do we want to include the updates need to the Infoblox integration as well in this PR or is there another PR for that one?

jdrew82 commented:
> Do we want to include the updates need to the Infoblox integration as well in this PR or is there another PR for that one?

Good catch! I dunno how I missed that one before.

glennmatthews commented on nautobot_ssot/migrations/0007_replace_dashed_custom_fields.py:
```diff
@@ -0,0 +1,51 @@
+        model = apps.get_model(ct.app_label, ct.model)
+        cf_list = []
+        for instance in model.objects.all():
+            new_custom_field_data = {}
+            for cf in relevant_custom_fields:
+                new_custom_field_data[cf.key] = instance._custom_field_data.pop(CF_KEY_CHANGE_MAP.get(cf.key), None)
+            instance._custom_field_data = new_custom_field_data
```
Still looks to me like this will throw away every custom_field_data entry that isn't part of CF_KEY_CHANGE_MAP... perhaps try and test the below?

```suggestion
            new_custom_field_data = instance._custom_field_data.copy()
            for cf in relevant_custom_fields:
                if cf.key not in new_custom_field_data:
                    new_custom_field_data[cf.key] = new_custom_field_data.pop(CF_KEY_CHANGE_MAP.get(cf.key), None)
            instance._custom_field_data = new_custom_field_data
```

jdrew82 replied:
Updated to match.

glennmatthews commented on nautobot_ssot/integrations/ipfabric/signals.py:
```diff
@@ -76,6 +76,7 @@ def nautobot_database_ready_callback(sender, *, apps, **kwargs):  # pylint: disa
-    create_custom_field("ssot-synced-from-ipfabric", "Last synced from IPFabric on", synced_from_models, apps=apps)
-    create_custom_field("ipfabric-site-id", "IPFabric Location ID", [Location], apps=apps, cf_type="type_text")
+    create_custom_field("system_of_record", "System of Record", synced_from_models, apps=apps, cf_type="type_text")
+    create_custom_field("last_synced_from_ssot", "Last sync from System of Record", synced_from_models, apps=apps)
```
nit: perhaps `last_synced_from_sor` would be more self-consistent?

jdrew82 replied:
Done.

jdrew82 commented on nautobot_ssot/integrations/ipfabric/constants.py:
```diff
@@ -22,3 +22,4 @@
 SAFE_DELETE_LOCATION_STATUS = CONFIG.get("ipfabric_safe_delete_location_status", "Decommissioning")
 SAFE_DELETE_VLAN_STATUS = CONFIG.get("ipfabric_safe_delete_vlan_status", "Deprecated")
 SAFE_DELETE_IPADDRESS_STATUS = CONFIG.get("ipfabric_safe_delete_ipaddress_status", "Deprecated")
+LAST_SYNCHRONIZED_CF_NAME = "last_synced_from_ssot"
```
IMO I think it should be something that we should utilize across the integrations as it's been extremely helpful with client engagements being able to determine which SoR some object came from when you have multiple SSoTs operating together. This might be moot point though with some of the changes coming in core Nautobot though?

jdrew82 commented:
Migration should now be updated to work accordingly:

- Look for any CustomFields that have a key with dashes that needs to be updated.
- Updates the key for those CustomFields to the new key and saves it.
- Looks through all ContentTypes that are related to those CustomFields and copies the value for the old CustomField to the new one.
- Deletes any of the old CustomFields that might remain.

glennmatthews commented on nautobot_ssot/migrations/0007_replace_dashed_custom_fields.py:
```diff
@@ -0,0 +1,53 @@
+from nautobot.extras.utils import FeatureQuery
+
+CF_KEY_CHANGE_MAP = {
+    "servicenow_last_synchronized": "ssot-synced-to-servicenow",
+    "last_synced_from_sor": "ssot-synced-from-ipfabric",
+    "ipfabric_site_id": "ipfabric-site-id",
+    "ssot_synced_to_infoblox": "ssot-synced-to-infoblox",
```
Sorry to keep nitpicking, but would it make sense to change `servicenow_last_synchronized` to `ssot_synced_to_servicenow` to be consistent with `ssot_synced_to_infoblox`?

jdrew82 replied:
Done.

glennmatthews commented on nautobot_ssot/migrations/0007_replace_dashed_custom_fields.py:
```diff
@@ -0,0 +1,53 @@
+                custom_field.key = new_key
+                custom_field.save()
+
+    for ct in ContentType.objects.filter(FeatureQuery("custom_fields").get_query()):
+        model = apps.get_model(ct.app_label, ct.model)
+        cf_list = []
+        for instance in model.objects.all():
```
Concerned about the performance overhead here of iterating through literally every instance of every model that supports any custom fields. Can we at least narrow this down to just the set of models that the old CFs from this app would exist for - looks like that'd be `[Device, DeviceType, Interface, Manufacturer, Location, VLAN, Role, IPAddress]`?

jdrew82 replied:
Done.

glennmatthews commented on nautobot_ssot/migrations/0007_replace_dashed_custom_fields.py:
```diff
@@ -0,0 +1,53 @@
+        cf_list = []
+        for instance in model.objects.all():
+            for new_cf, old_cf in CF_KEY_CHANGE_MAP.items():
+                if old_cf in instance._custom_field_data and new_cf in instance._custom_field_data:
```
Should this be?

```suggestion
                if old_cf in instance._custom_field_data and new_cf not in instance._custom_field_data:
```

?

jdrew82 replied:
Nope, I do this check to ensure that the old and new CustomField keys are in place for the copy to occur. The pop should remove the old one when it copies the data to the new CustomField key and then removes the CustomField at the end.

jdrew82 commented:
I've also added a fix for the Device42 integration as it does create CustomFields that could potentially have spaces in the slug/key. I've corrected it to replace spaces and dashes with underscores.

jeffkala commented:
Any ETA for this? @jdrew82

jdrew82 commented:
> Any ETA for this? @jdrew82

Just asked @glennmatthews for another review of the PR. After that we should be able to merge it and cut a new release.

glennmatthews commented on nautobot_ssot/migrations/0007_replace_dashed_custom_fields.py:
```diff
@@ -0,0 +1,58 @@
+                custom_field.key = new_key
+                custom_field.save()
+
+    for ct in ContentType.objects.filter(FeatureQuery("custom_fields").get_query()):
```
Doesn't look like `ct` is actually used anywhere - unnecessary outer `for` loop?

jdrew82 replied:
Good catch. You're right, we're not using that anymore so removed.

glennmatthews commented on nautobot_ssot/tests/device42/unit/test_utils_nautobot.py:
```diff
@@ -153,13 +153,13 @@ def test_update_custom_fields_add_cf(self):
-        self.assertEqual(len(test_site.get_custom_fields()), 3)
+        self.assertEqual(len(test_site.get_custom_fields()), 4)
         mock_cfs = {
             "Test Custom Field": {"key": "Test Custom Field", "value": None, "notes": None},
```
Perhaps?

```suggestion
            "Test Custom Field": {"key": "Test_Custom_Field", "value": None, "notes": None},
```

jdrew82 replied:
This part of the test is actually what the information from Device42 looks like and isn't the key for the CustomField like you might think. The name is admittedly a little confusing.

jdrew82 commented on nautobot_ssot/migrations/0007_replace_dashed_custom_fields.py:
```diff
@@ -0,0 +1,58 @@
+    for new_key, old_key in CF_KEY_CHANGE_MAP.items():
+        if not CustomField.objects.filter(key=new_key).exists():
+            for custom_field in CustomField.objects.filter(key=old_key):
+                original_key = custom_field.key
+                print(
+                    f'   CustomField instance "{custom_field.label}" key attribute "{original_key}" is being changed to "{new_key}".'
+                )
```
Good point. Fixed.

jdrew82 commented on nautobot_ssot/integrations/ipfabric/utilities/nbutils.py:
```diff
@@ -229,10 +230,10 @@ def _tag_object(nautobot_object):
-                custom_field_obj, _ = CustomField.objects.get(
-                    key="ssot-synced-from-ipfabric",
+            # Ensure that the LAST_SYNCHRONIZED_CF_NAME custom field is present
+            if not any(cfield for cfield in CustomField.objects.all() if cfield.key == LAST_SYNCHRONIZED_CF_NAME):
+                custom_field_obj = CustomField.objects.get(
+                    key=LAST_SYNCHRONIZED_CF_NAME,
                 )
```
Good question. That part of the code predates this PR and appears to come from the 2.0 update. I've gone ahead and made the suggested change though as I do agree it's more efficient.

jdrew82 replied:
I've removed it as it is no longer necessary with the CustomField being created via signal.
